### PR TITLE
feat: Redis as (distributed) cache

### LINKED ISSUE #17: Implement support for Radis as an alternative to the available in-memory cache

### PR DESCRIPTION
**Pull request by tk-innoq**
## Related issue(s)

closes #17 

## Checklist

- [x] I agree to follow this project's [Code of Conduct](../CODE_OF_CONDUCT.md).
- [x] I have read, and I am following this repository's [Contributing Guidelines](../CONTRIBUTING.md).
- [x] I have read the [Security Policy](../SECURITY.md).
- [x] I have referenced an issue describing the bug/feature request.
- [x] I have added tests that prove the correctness of my implementation.
- [x] I have updated the documentation.

## Description

Before this PR, there was just the in-memory cache. This PR makes cache configurable by introducing a `cache` property on the top level of heimdall configuration, like shown below:

```yaml
cache:
  type: in-memory # or noop, redis, redis-cluster, or redis-sentinel
  config: # depends on the type
    # type specific configuration goes here
```

As also shown in the snipped above, there are a couple of cache types one can now configure:
* `noop` - Caching is disabled completely. That means any cache settings on mechanism level do not have any effect. No configuration is supported for this type.
* `in-memory` - This is the default setting, which implements the behavior used by heimdall so far. No configuration is supported for this type.
* `redis` - A Redis single instance can be used as a cache store. Configuration is required (see below).
* `redis-cluster`  - A Redis cluster can be used as a distributed cache store. Configuration is required (see below).
* `redis-sentinel` -  A Redis sentinel can be used as a distributed cache store. Configuration is required (see below)

Common configuration options for all redis cache types
```yaml
# Optional credentials to connect to Redis. Either 'path' or 'username` and `password` can be used.
credentials:
  # Optional path to the credentials file
  path: /path/to/credentials/file
  # Optional username to use to connect to the Redis DB/Cluster
  username: foo
  # Mandatory password to use to connect to the Redis DB/Cluster
  password: foobar

# Alternatively to specifying a map for the credentials property,
# one can reference a file with credentals: /path/to/file with username
# and password configured. This way you can rotate these settings
# without the need to restart heimdall.

# Optional configuration for the client cache to reduce roundtrip time
client_cache:
  # Whether the client cache should be disabled. Defaults to false
  disabled: false
  # Specifies the maximum TTL on the client side for all used keys. 
  # If the key's TTL on the server is smaller than the client side TTL, 
  # the client side TTL will be capped.
  ttl: 5m
  # Optional client side cache size that bind to each TCP connection
  # to a single redis instance. Defaults to 128MB
  size_per_connection: 128MB 
  
# Optional connection buffer limits
buffer_limit: 
  # Read size for each connection. Defaults to 0.5MB
  read: 1MB 
  # Write size for each connection. Defaults to 0.5MB 
  write: 1MB
  
# Optional connection time outs
timeout:
  # The write timeout for connections. Defaults to 150s
  write: 150s

# Optional data flush delay. When greater than zero, pauses pipeline write loop
# for some time (not larger than max_flush_delay)  after each flushing of data to
# the connection. This gives pipeline a chance to collect more commands to 
# send to Redis. Adding this delay increases latency, reduces throughput, but in
# most cases may significantly reduce application and Redis CPU utilization due
# to less executed system calls. By default, the data is flushed to the connection
# without extra delays. Depending on network latency and application-specific
# conditions the value of max_flush_delay may vary, something like 20 µs should
# not affect latency/throughput a lot but still produce notable CPU usage
# reduction under load. Defaults to 0s
max_flush_delay: 20us # 20µs

# Optional TLS configuration. TLS is enabled by default
# The system trust store is used to verify the TLS server certificates
tls:
  # Disables TLS. Defaults to false
  disabled: false

  # Further properties as already known from service configurations. All are optional
  # If client authentication is required, configure the key_store
  key_store:
    path: /path/to/keystore/file.pem
    password: VerySecret!
  # Which key/certificate to use from the key store. If not set, the first one is taken
  key_id: foo
  # Min TLS version. Defaults to TLS1.3
  min_version: TLS1.2
  # Which cipher suites to use with TLS1.2 (all supported cipher suites are shown below)
  cipher_suites:
    - "TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256"
    - "TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256"
    - "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256"
    - "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256"
    - "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384"
    - "TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384"
    - "TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256"
    - "TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256"
```

Additional configuration properties per redis type:
* `redis`:
   ```yaml
   # Mandatory Redis instance address to connect to.
   address: foo:12345
   # Optional Redis DB ID to use. Defaults to 0
   db: 0
   ```
* `redis-cluster`:
   ```yaml
   # Mandatory list of Redis nodes to connect to
   nodes:
     - foo:1234
     - bar:1234
   ```
* `redis-sentinel`:
   ```yaml
   # Optional Redis DB ID to use. Defaults to 0
   db: 0
   # Mandatory sentinel master set name
   master: whatever
   # Mandatory list of Redis nodes to connect to
   nodes:
     - foo:1234
     - bar:1234
   ```

## Additional Information

* If Redis (Standalone/Cluster/sentinel) is not available at start up, heimdall will refuse to start. At run time however, only warnings will be emitted if communication to Redis is not possible, but the functionality will be preserved with, depending on your rules, some performance degradation. Corresponding metrics are exposed as well.
* Same credentials and TLS settings are used for sentinel and master nodes connections if sentinel is configured.
* Values to be cached may contain highly sensible information. For that reason, TLS is enabled by default (see configuration options above). You should however also ensure your Redis setup has data encryption configured as well.
* As shown in the configuration snippet above, this PR implements support for short living DB credentials. To make use of that, configure a valid path to the corresponding yaml/json file with `username` and `password` managed externally. 
   Example:
   
   ```yaml
   # Redis cache config
   credentials: 
    path: /path/to/credentials.yaml
   ```
   And the contents of the `credentials.yaml` from above would look like
   
   ```yaml
   username: foo
   password: foobar
   ```

   In addition, you have to allow reloading of externally managed secrets by setting `secrets_reload_enabled` to `true` on the top level of heimdall configuration.
   Example:
   
   ```yaml
   serve:
     # your serve settings
   
   cache:
     # your cache settings (see above)
   
   secrets_reload_enabled: true
   ```

## Current Status of the PR

Fully functional. Documentation needs to be updated.

### CODE REVIEW

**Comment by tk-innoq** [comment]
Relates to: https://github.com/dadrus/heimdall/issues/17

**Comment by dadrus** [comment]
@tk-innoq: Thank you very much for the PR. Would you please fix the style and test issues? I'll take a closer look at the PR as soon as this is done.

File: internal/cache/module.go [hunk 1]
```diff
@@ -35,14 +36,35 @@
...
+	case "disabled":
+		logger.Info().Msg("Cache is disabled")
+		cache = noopCache{}
+	default:
+		logger.Info().Msg("Fallback: noop cache")
+		cache = noopCache{}
 	}
 
-	logger.Info().Msg("Cache is disabled")
+	if cache == nil {
```
[the remark this answers is not in the document]
Comment (tk-innoq) [reply internal/cache/module.go]: Fixed

**Comment by dadrus** [comment]
@tk-innoq: After taking a closer look, it doesn't look like to PR is ready. What are your plans related to this PR?

**Comment by dadrus** [comment]
Another question is related to the codeQL alerts above: What was actually your goal with the new spans? I mean, the redisotel does actually the relevant stuff. No?

**Comment by dadrus** [comment]
The PR is complete from functionality point of view. Waiting for finalization of #1075 to add the new cache related content to

**Pull request merged**